OSPF: support detailed neighbor changes log and authentication on VLAN interfaces

c-po opened the pull request:
I've added two new nodes to the data model:

* Possibility to extend the default-on `log-adjacency-changes` statement with the `detail` option
* VLAN interfaces gained the possibility to specify the OSPF message-digest authentication key(s)

@carlbuchmann @aphillipps if you have any suggestion to improve the data model I'm happy to change it!

Commit:
ospf: provide option to enable "log-adjacency-changes detail"

Commit:
ospf: support setting md5 key on VLAN interfaces

carlbuchmann commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -76,6 +76,12 @@ interface {{ vlan_interface }}
 {%         endif %}
+{%         if vlan_interfaces[vlan_interface].ospf_message_digest is defined and vlan_interfaces[vlan_interface].ospf_message_digest is not none %}
+   ip ospf authentication message-digest
+{%           for id in vlan_interfaces[vlan_interface].ospf_message_digest %}
```
Please sort the id as per below, to match cli output when multiple ids are provided.
{% for id in vlan_interfaces[vlan_interface].ospf_message_digest | sort %}

c-po replied:
Updates as requested.

carlbuchmann commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
```diff
@@ -1264,6 +1267,7 @@ router_ospf:
     < process_id >:
       passive_interface_default: < true | false >
       router_id: < IPv4_address >
+      log_detailed_adjacency_changes:
```
update key to match the cli syntax to avoid confusion.
`log_adjacency_changes_detail: < true | false >`

c-po replied:
Updates as requested.

carlbuchmann commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/router-ospf.j2:
```diff
@@ -7,6 +7,9 @@ router ospf {{ process_id }} vrf {{ router_ospf.process_ids[process_id].vrf }}
 {%         else %}
 router ospf {{ process_id }}
 {%         endif %}
+{%         if router_ospf.process_ids[process_id].log_detailed_adjacency_changes is defined and router_ospf.process_ids[process_id].log_detailed_adjacency_changes == true %}
```
update as per data model change request.

c-po replied:
Updates as requested.

carlbuchmann reviewed (approved):
Thanks for the changes LGTM

aphillipps reviewed (commented):
I would change it to the following as you could have your keys configured and still be in simple authentication and switch to message-digest authentication later:
```yaml
    ospf_authentication: < none | simple | message-digest >
    ospf_authentication_key: < encrypted_key >
    ospf_message_digest_key:
         < id >:
             method: < md5 | sha1 | sha 256 | sha384 | sha512 >
             key: "< encrypted_password >"
```

c-po commented:
That would work out, too and be okay for me.

The only thing I would not add is `method`, as currently only `md5` is supported (according to my vEOS-LAB unit). Is there other support on other Arista gear?

carlbuchmann commented:
@c-po 
> The only thing I would not add is `method`, as currently only `md5` is supported (according to my vEOS-LAB unit). Is there other support on other Arista gear?

Yes, several methods are available to be configured, that was a good catch by @aphillipps !

The following methods are available on Arista -> output from vEOS-LAB 4.23.4M:
```
DC1-BL1A(config-if-Vl99)#ip ospf message-digest-key 1 ?
  md5     Configure MD5 authentication
  sha1    Configure SHA1 authentication
  sha256  Configure SHA256 authentication
  sha384  Configure SHA384 authentication
  sha512  Configure SHA512 authentication
```

carlbuchmann commented:
@c-po and @aphillipps 

Note: this doesn't mean these methods are "supported" on vEOS-LAB, but if they are present as options in the CLI is may eventually be supported.

Also, should we call the key: `authentication_method:` as opposed to `method:`

c-po commented:
@aphillipps thank your for your input - I've updated this PR accordingly.

Commit:
README: unify coding style on encrypted_password - always use quotes

aphillipps commented:
After discussion, in order to standardize the keys across AVD, we will use the key "hash_algorithm" every time it is needed. So, please change it according by replacing the key "method" by the key "hash_algorithm". We will update the current snmp template with the same key as well. As I had no indentation in my first comment, it was maybe not clear, but configuration of the ospf key should not depend on the configuration of the ospf authentication type (simple or message-digest). Indeed, you can have keys configured for both types (simple and message-digest) on the same time and then switch from one authentication to another. So, please review your jinja template. Thanks.

Commit:
ospf: rename "method" to "hash_algorithm" as per new policy

The key name in the datamodel should be standardised in AVD. If a hash is used
at any location the key used for selecting the has algorithm should always be:
hash_algorithm.

c-po commented:
Hi @aphillipps, thats not a problem. I know designing data model is non trivial as they will stick around forever and migration is hard. So it should be right from the beginning.

The revised implementation would render a snipped like this:

```
interface Vlan3196
   description CROSSLINK
   ip address 10.0.130.197/30
   ip ospf network point-to-point
   ip ospf area 0.0.0.0
   ip ospf authentication message-digest
   ip ospf authentication-key 7 foo
   ip ospf message-digest-key 1 md5 7 bar
   ip ospf message-digest-key 2 sha1 7 baz
```

I guess this is what you requested. All keys are present in the config but the selection is only done by `ip ospf authentication message-digest`. Did I get you right?

Commit:
ospf: always render all possible authentication keys to config

Key selection is done exclusively by "ip ospf authentication [message-digest]"

Commit:
ospf: ethernet: interfaces: add authentication support

aphillipps commented:
Hello @c-po, thanks for your contribution and sorry for the delay. I have changed a bit your template to comply with our guidelines. As there are not published yet, you could not guess them ;-) We will publish them soon to avoid such issues. I have also corrected my initial suggested data-model to comply to these guidelines. See below. Basically when there are several values possible, the key should be plural, so I changed ospf_message_digest_key to ospf_message_digest_keys. In the FOR loop we will then reference the singular form of that key meaning here ospf_message_digest_key. There was also a little typo on your side but no worries. Thanks a lot for your patience and contribution. I let you test the new format of the template and get back to me and then we will hopefully merged it.

```yaml
    ospf_authentication: < none | simple | message-digest >
    ospf_authentication_key: < encrypted_key >
    ospf_message_digest_keys:
         < id >:
             hash_algorithm: < md5 | sha1 | sha 256 | sha384 | sha512 >
             key: "< encrypted_password >"
```

```jinja2
{%             if ethernet_interfaces[ethernet_interface].ospf_area is defined and ethernet_interfaces[ethernet_interface].ospf_area is not none %}
   ip ospf area {{ ethernet_interfaces[ethernet_interface].ospf_area }}
{%             endif %}
{%             if ethernet_interfaces[ethernet_interface].ospf_authentication is defined and ethernet_interfaces[ethernet_interface].ospf_authentication is not none %}
{%                 if ethernet_interfaces[ethernet_interface].ospf_authentication == "simple" %}
   ip ospf authentication
{%                 elif ethernet_interfaces[ethernet_interface].ospf_authentication == "message-digest" %}
   ip ospf authentication message-digest
{%                 endif %}
{%             endif %}
{%             if ethernet_interfaces[ethernet_interface].ospf_authentication_key is defined and ethernet_interfaces[ethernet_interface].ospf_authentication_key is not none %}
   ip ospf authentication-key 7 {{ ethernet_interfaces[ethernet_interface].ospf_authentication_key }}
{%             endif %}
{%             if ethernet_interfaces[ethernet_interface].ospf_message_digest_keys is defined and ethernet_interfaces[ethernet_interface].ospf_message_digest_keys is not none %}
{%               for ospf_message_digest_key in ethernet_interfaces[ethernet_interface].ospf_message_digest_keys | arista.avd.natural_sort %}
   ip ospf message-digest-key {{ ospf_message_digest_key }} {{ ethernet_interfaces[ethernet_interface].ospf_message_digest_keys[ospf_message_digest_key].hash_algorithm }} 7 {{ ethernet_interfaces[ethernet_interface].ospf_message_digest_keys[ospf_message_digest_key].key }}
{%               endfor %}
{%             endif %}
```

Commit:
ospf: adjust to new coding guidelines (use key plural)

c-po commented:
@aphillipps not a problem - there you go!

aphillipps commented:
LGTM

carlbuchmann reviewed (approved):
LGTM
